src/module.ts

```diff
@@ -1,56 +1,71 @@
 
-export type { Rule }
+export interface ModuleHooks {
+  'robots:generate:before': (moduleOptions: ModuleOptions) => void
+  'robots:generate:done': (content: string) => void
+}
 
-export type ModuleOptions = Rule | Rule[] | (() => Rule | Rule[])
+export default defineNuxtModule({
+  meta: {
+    name,
+    version,
+    configKey: 'robots'
+  },
 
-async function getOptions (moduleOptions: ModuleOptions): Promise<Rule[]> {
-  if (typeof moduleOptions === 'function') {
-    moduleOptions = await moduleOptions.call(this)
-  }
+  setup (moduleOptions, nuxt) {
+    if (!Array.isArray(moduleOptions)) {
+      moduleOptions = [{
+        UserAgent: '*',
+        Disallow: '',
+        ...moduleOptions
+      }]
+    }
 
-  if (Array.isArray(moduleOptions)) {
-    return moduleOptions
-  }
+    let staticRules: RuleInterface[] = []
 
-  let { robots } = this.options
+    // Loading existing robots.txt
+    nuxt.hook('build:before', async () => {
+      const staticDir = isNuxt2() ? nuxt.options.dir.static : nuxt.options.dir.public
+      const staticFilePath = resolve(nuxt.options.srcDir, staticDir, 'robots.txt')
 
-  if (typeof robots === 'function') {
-    robots = await robots.call(this)
-  }
+      if (existsSync(staticFilePath)) {
+        const content = readFileSync(staticFilePath).toString()
 
-  if (Array.isArray(robots)) {
-    return robots
-  }
+        staticRules = await getRules.call(nuxt, parseFile(content))
+      }
+    })
 
-  return [{
-    UserAgent: '*',
-    Disallow: '',
-    ...robots,
-    ...moduleOptions
-  }]
-}
+    // Generate robots.txt
+    nuxt.hook('generate:done', async () => {
+      // @ts-ignore
+      await nuxt.callHook('robots:generate:before', moduleOptions)
 
-const nuxtModule: Module<ModuleOptions> = async function (moduleOptions) {
-  const options: Rule[] = await getOptions.call(this, moduleOptions)
+      const generateFilePath = resolve(nuxt.options.rootDir, nuxt.options.generate.dir, 'robots.txt')
+      const rules: RuleInterface[] = await getRules.call(nuxt, moduleOptions)
+      const content = render([...staticRules, ...rules])
 
-  build.bind(this)()
-  generate.bind(this)(options)
-  middleware.bind(this)(options)
-}
+      writeFileSync(generateFilePath, content)
 
-;(nuxtModule as any).meta = { name, version }
+      // @ts-ignore
+      await nuxt.callHook('robots:generate:done', content)
+    })
 
-declare module '@nuxt/types' {
-  interface NuxtConfig { [CONFIG_KEY]?: ModuleOptions } // Nuxt 2.14+
-  interface Configuration { [CONFIG_KEY]?: ModuleOptions } // Nuxt 2.9 - 2.13
-}
+    // Middleware /robots.txt
+    nuxt.options.serverMiddleware.unshift({
+      path: '/robots.txt',
+      async handler (req, res) {
+        const rules: RuleInterface[] = await getRules.call(nuxt, moduleOptions, req)
 
-export default nuxtModule
+        res.setHeader('Content-Type', 'text/plain')
+        res.end(render([...staticRules, ...rules]))
+      }
+    })
```

Comment:
This will only run in dev mode, so we either need to add a static `robots.txt` to the output public directory or add a server-middleware _file_ which can render a dynamic `robots.txt` on per-request.

Reply:
But isn't that an serverMiddleware?

Reply:
Does it need to be a file? With the function is legacy, is that it?

Reply:
@danielroe any idea how i can share module options with middleware file?

Reply:
you can set the runtime configuration (`nuxt.options.runtimeConfig`), which your middleware file can access with `useRuntimeConfig`. Would that work?

Reply:
In this case it is necessary to serialize functions

Reply:
Then you should not use runtimeConfig.

This summarises a couple of the options: https://github.com/nuxt/framework/issues/3845#issuecomment-1098938345

Reply:
@danielroe  Not working  https://stackblitz.com/edit/github-bsr21s-bf23tt